designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+be captured by reference.  Automatic variables are those without
+explicit capture declarations; all variables in this proposal for Stan
+will behave as automatic variables.
+
+#### R
+
+In R, the expression `function(u) { return(1 / (1 + exp(-u))) }`
+defines a function that denotes the inverse logit function.  It can be
+assigned to a variable, e.g.,
+
+```
+inv_logit <- function(u) { return(1 / (1 + exp(-u))); }
+theta <- inv_logit(0.2)
+```
+
+Unlike C++ or Stan, R requires the outer parentheses for returns.  R
+provides a convenient shortcut that the body of a function returns its
+last evaluated expression if there is no return statement, so the
+inverse logit function can be rewritten as
+
+```
+inv_logit <- function(u) { 1 / (1 + exp(-u)) }
+```
+
+We can go one step further and drop the braces, as they are only there
+to allow a sequence of statements to be grouped,
+
+```
+inv_logit <- function(u) 1 / (1 + exp(-u))
+```
+
+This is the abbreviated syntax we recommend here for Stan lambdas,
+where the equivalent would be
+
+```
+real(real) ilogit = (real u) { return 1 / (1 + exp(-u)); };
+```
+
+or in abbreviated form,
+
+```
+real(real) ilogit = (real u) 1 / (1 + exp(-u));
+```
+
+R uses the unusual mechanism of dynamic lexical scoping, meaning that
+a lambda will capture whichever variable exists in its environment
+when executed.  This can lead to non-deterministic behavior of
+variable scopes at runtime.  This proposal for Stan is to use the more
+traditional approach of static lexical scoping, which is what is used
+in C++.
+
+#### Python
+
+Python allows lambdas such as `lambda u : 1 / (1 + exp(-u))`.
+Multiple argument functions my be `lambda x, y : (x**2 + y**2)**0.5`.
+Python captures variables by reference in lambdas, as the following
+example demonstrates.
+
+```
+>>> x = 3
+>>> f = lambda y : (x**2 + y**2)**0.5
+>>> f(4)
+5.0
+>>> x = 10
+>>> f(4)
+10.770329614269007
+```
+
+This is the same capture-by reference that is used by R and this
+proposal for Stan, as well as the behavior for C++ if the captures
+specification takes all implicit captures by reference (`[&]`).
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1.  Will Stan permit potentially risky capture of local variables and
+    function arguments?  If it is allowed (as it presumably will be,
+    as it's very very limiting not to) is there a way to flag cases
+    that might be potentially dangerous and provide warnings?  
+
+2.  What should the syntax look like?  This proposal just follows the
+	C++ syntax for both types and closures.
+
+3.  Should variables be captured by value instead of reference? That
+    avoids them changing later and avoids dangling references, but it
+    also requires copies and is counterintuitive for those used to
+    R or Python.
+
+4.  Should full covariance and contravariance for function and array
+    types be required for assignment (including function calls)?
```

Comment:
It would be very nice, but it'd be a lot of work to fix the whole language.

Reply:
We don't have to consider this as part of the closures proposal, I think :) But I'm curious (and maybe we should take this to a github issue) - what all would be required? I guess if C++ didn't support automated casting we'd have to generate code that did the casting?